### PR TITLE
improve type assertion error

### PR DESCRIPTION
**Pull request by mengqiy**
Per discussion https://github.com/kubernetes/kubernetes/pull/45980#issuecomment-302524829.

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/util/mergepatch/errors.go [hunk 1]
```diff
@@ -35,6 +36,12 @@ func ErrBadArgType(expected, actual string) error {
 	return fmt.Errorf("expected a %s, but received a %s", expected, actual)
 }
 
+func ErrBadArgTypeAssertion(expected, actual interface{}) error {
```
Comment (apelisse) [inline staging/src/k8s.io/apimachinery/pkg/util/mergepatch/errors.go]: Should you remove `ErrBadArgType`?

Comment (mengqiy) [inline staging/src/k8s.io/apimachinery/pkg/util/mergepatch/errors.go]: There are some usage left. e.g. https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go#L571.
WDYT leave them as is?

Comment (apelisse) [inline staging/src/k8s.io/apimachinery/pkg/util/mergepatch/errors.go]: I don't know ... The next person coming won't know which one to use :-/

Comment (mengqiy) [inline staging/src/k8s.io/apimachinery/pkg/util/mergepatch/errors.go]: Added a comment for that.

Comment (apelisse) [inline staging/src/k8s.io/apimachinery/pkg/util/mergepatch/errors.go]: Wait, there are only a couple left? Can't you just change that couple and remove the function once and for all?

Comment (apelisse) [inline staging/src/k8s.io/apimachinery/pkg/util/mergepatch/errors.go]: Looks like the one with Nil can't be replaced easily, but maybe the error isn't super correct either anyway.

**Comment by mengqiy** [comment]
@apelisse PTAL

File: staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go [hunk 2]
```diff
@@ -568,7 +567,7 @@ func StrategicMergeMapPatch(original, patch JSONMap, dataStruct interface{}) (JS
 
 func getTagStructType(dataStruct interface{}) (reflect.Type, error) {
 	if dataStruct == nil {
-		return nil, mergepatch.ErrBadArgType("struct", "nil")
+		return nil, mergepatch.ErrBadArgType(struct{}{}, nil)
```
Comment (apelisse) [inline staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go]: That is somewhat weird because nil is not a type, but yeah, that works.

Comment (caesarxuchao) [inline staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go]: Why being nil is a `ErrBadArgType`?

File: staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go [hunk 3]
```diff
@@ -577,7 +576,7 @@ func getTagStructType(dataStruct interface{}) (reflect.Type, error) {
 	}
 
 	if t.Kind() != reflect.Struct {
-		return nil, mergepatch.ErrBadArgType("struct", t.Kind().String())
+		return nil, mergepatch.ErrBadArgType(struct{}{}, dataStruct)
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go]: Should be a `ErrBadArgKind`, right? Type and Kind are different.

**Comment by mengqiy** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by mengqiy** [comment]
We should distinguish type and kind.
I updated the code.
@caesarxuchao PTAL

File: staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go [hunk 4]
```diff
@@ -505,8 +505,7 @@ func diffListsOfMaps(original, modified []interface{}, t reflect.Type, mergeKey
 func getMapAndMergeKeyValueByIndex(index int, mergeKey string, listOfMaps []interface{}) (map[string]interface{}, interface{}, error) {
 	m, ok := listOfMaps[index].(map[string]interface{})
 	if !ok {
-		t := reflect.TypeOf(listOfMaps[index])
-		return nil, nil, mergepatch.ErrBadArgType("map[string]interface{}", t.Kind().String())
+		return nil, nil, mergepatch.ErrBadArgType(m, listOfMaps[index])
```
Comment (apelisse) [inline staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go]: Are you sure this is a bad type?

Comment (mengqiy) [inline staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go]: I think so.
If `listOfMaps[index]` is `map[int]interface{}`, but not `map[string]interface{}`, this is bad type, even though their kinds are both `Map`.

Comment (apelisse) [inline staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go]: Good point, thanks

**Comment by mengqiy** [comment]
@k8s-bot unit test this

**Pull request merged**